### ISSUE TITLE
Testing with MonoGame : threading error

### ISSUE THREAD

**Issue by spaaaaam** [MonoGame/MonoGame#6413]
Hello,

I am currently trying to add some visual tests to my game. In managed to write working tests when running independently, but I get an exception when I run all the tests together.

Here's a minimal example that exhibits the problem :
```
    [TestFixture]
    [Apartment(ApartmentState.STA)]
    public class SimplestGame
    {
        TimeSpan timeLimit = new TimeSpan(0, 0, 0, 0, 100);

        [Test]
        public void RunSimplestGame()
        {
            Console.WriteLine("Starting...");
            var game = new DummyGame(timeLimit);
            Console.WriteLine("Running...");
            game.Run();
            Console.WriteLine("Exiting...");
            game.Dispose();
        }
    }

    [TestFixture]
    [Apartment(ApartmentState.STA)]
    public class SimplestGame2
    {
        TimeSpan timeLimit = new TimeSpan(0, 0, 0, 0, 100);

        [Test]
        public void RunSimplestGame()
        {
            Console.WriteLine("Starting...");
            var game = new DummyGame(timeLimit);
            Console.WriteLine("Running...");
            game.Run();
            Console.WriteLine("Exiting...");
            game.Dispose();
        }
    }

    public class DummyGame : Game
    {
        private readonly GraphicsDeviceManager graphics;
        private readonly TimeSpan limit;
        private SpriteBatch spriteBatch;

        public DummyGame(TimeSpan limit)
        {
            graphics = new GraphicsDeviceManager(this);
            this.limit = limit;
        }

        protected override void Initialize()
        {
            base.Initialize();
        }

        protected override void LoadContent()
        {
            Content.RootDirectory = "Content";
            spriteBatch = new SpriteBatch(GraphicsDevice);
        }

        protected override void UnloadContent()
        {
        }

        protected override void Update(GameTime gameTime)
        {
            if (GamePad.GetState(PlayerIndex.One).Buttons.Back == ButtonState.Pressed
                || Keyboard.GetState().IsKeyDown(Keys.Escape)
                || Elapsed(gameTime))
            {
                Exit();
            }
            base.Update(gameTime);
        }

        protected override void Draw(GameTime gameTime)
        {
            GraphicsDevice.Clear(Color.CornflowerBlue);
            base.Draw(gameTime);
        }

        private bool Elapsed(GameTime gameTime)
        {
            if (gameTime.TotalGameTime > limit)
                return true;

            return false;
        }
    }
```
Both tests (they are identical) run fine if I run them alone, but the second one crashes if I run them together in nunit, with the following error :

```
One or more errors occurred.
Stack trace:
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <9bbab8f8a2a246e98480e70b0839fd67>:0 
  at System.Threading.Tasks.Task`1[TResult].GetResultCore (System.Boolean waitCompletionNotification) [0x0002b] in <9bbab8f8a2a246e98480e70b0839fd67>:0 
  at System.Threading.Tasks.Task`1[TResult].get_Result () [0x0000f] in <9bbab8f8a2a246e98480e70b0839fd67>:0 
  at MonoDevelop.UnitTesting.NUnit.NUnitAssemblyTestSuite.RunUnitTest (MonoDevelop.UnitTesting.UnitTest test, System.String suiteName, System.String pathName, System.String testName, MonoDevelop.UnitTesting.TestContext testContext) [0x00245] in <5a50c7f8fd15492b8f57b877d888bb04>:0 
  at MonoDevelop.UnitTesting.NUnit.NUnitTestSuite.OnRun (MonoDevelop.UnitTesting.TestContext testContext) [0x00001] in <5a50c7f8fd15492b8f57b877d888bb04>:0 
  at MonoDevelop.UnitTesting.UnitTest.Run (MonoDevelop.UnitTesting.TestContext testContext) [0x00020] in <14600d39865a4deabab34102d35bdc57>:0 
```

If I comment out the `` game.dispose()`` instruction in the ``RunSimplestGame`` method, I can run them together without error, so I suspect that something is shared between the two tests and disposed before the second one is finished.

Am I missing something? Any idea of what needs to be done to ensure that both tests are effectively independent?

Thanks

#### What version of MonoGame does the bug occur on:
- MonoGame 3.7.0.1644

#### What operating system are you using:
- Ubuntu 18.04

#### What MonoGame platform are you using:

- DesktopGL

**Comment by D4v1ks** [comment]
Run the tests in different processes.

https://docs.microsoft.com/pt-pt/windows/desktop/ProcThread/about-processes-and-threads

**Comment by spaaaaam** [comment]
Hi, 

sorry for the answering delay, but my computer died not long after posting this issue.

I am now on another PC (same setup, except ubuntu version is 16.04) on which I do not reproduce the problem.

Thanks for the suggestion anyway. Closing.

**Comment by spaaaaam** [comment]
Hello,

I stumbled upon the same problem on another machine. 

Following @D4v1ks advice I tried to run the tests with `--process=Multiple` nunit option, but without luck so far. If I understand correctly, this is because nunit run distinct assemblies in separate processes, but in my case, both tests are in the same assembly (see [this](https://stackoverflow.com/a/48326114)). 

Is there any other way to run individual tests in separate processes?

For the record, running the tests in the console gives some more details:

```
Errors, Failures and Warnings

1) Error : /home/.../bin/Debug/MyGame.Tests.dll
System.Runtime.Remoting.RemotingException : Tcp transport error.
  ----> System.Runtime.Remoting.RemotingException : Connection closed
--RemotingException


Server stack trace: 
  at System.Runtime.Remoting.Channels.Tcp.TcpMessageIO.ReceiveMessageStatus (System.IO.Stream networkStream, System.Byte[] buffer) [0x00017] in <981d2e773a85450680fa518175a3273b>:0 
  at System.Runtime.Remoting.Channels.Tcp.TcpClientTransportSink.ProcessMessage (System.Runtime.Remoting.Messaging.IMessage msg, System.Runtime.Remoting.Channels.ITransportHeaders requestHeaders, System.IO.Stream requestStream, System.Runtime.Remoting.Channels.ITransportHeaders& responseHeaders, System.IO.Stream& responseStream) [0x0005e] in <981d2e773a85450680fa518175a3273b>:0 
  at System.Runtime.Remoting.Channels.BinaryClientFormatterSink.SyncProcessMessage (System.Runtime.Remoting.Messaging.IMessage msg) [0x00066] in <981d2e773a85450680fa518175a3273b>:0 

Exception rethrown at [0]: 
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_remoting_wrapper(intptr,intptr)
  at (wrapper remoting-invoke) NUnit.Engine.Agents.RemoteTestAgent.Run(NUnit.Engine.ITestEventListener,NUnit.Engine.TestFilter)
  at NUnit.Engine.Runners.ProcessRunner.RunTests (NUnit.Engine.ITestEventListener listener, NUnit.Engine.TestFilter filter) [0x00025] in <a3ad5449913a4202999529a949a30f07>:0 
--RemotingException
  at System.Runtime.Remoting.Channels.Tcp.TcpMessageIO.StreamRead (System.IO.Stream networkStream, System.Byte[] buffer, System.Int32 count) [0x00011] in <981d2e773a85450680fa518175a3273b>:0 
  at System.Runtime.Remoting.Channels.Tcp.TcpMessageIO.ReceiveMessageStatus (System.IO.Stream networkStream, System.Byte[] buffer) [0x00000] in <981d2e773a85450680fa518175a3273b>:0 

```

Any help would be greatly appreciated!

**Comment by cra0zy** [comment]
Gonna close this as it is neither an issue or feature request.

**Comment by cra0zy** [comment]
I don't know how to setup this, not my area of expertise, tho this type of questions should be asked on the community site and not on the issue tracker: http://community.monogame.net

**Comment by spaaaaam** [comment]
Ok, thanks. I put it there: http://community.monogame.net/t/testing-with-monogame-on-linux-threading-error/10991

**Comment by spaaaaam** [comment]
Hello,

I recently tried to give another go at this problem with MG3.8. First I had to remove the `[Apartment(ApartmentState.STA)]` attributes which are not supported anymore on Linux in the code above (`Apartment state cannot be set on this platform.`).

Each test runs fine separately. When launching the two together, I now get a segmentation fault : `Exit code is 139`.

Debugging leads me to the [`DisposeContexts` static method `in GraphicsDevice.OpenGL.cs`](https://github.com/MonoGame/MonoGame/blob/e360e21714921cfe98a4f90aaf8e2573dae0191f/MonoGame.Framework/Platform/Graphics/GraphicsDevice.OpenGL.cs#L518).

If I jump over the `Sdl.GL.DeleteContext(_disposeContexts[i]);` the tests run fine. So apparently deleting the context triggers an access to an unauthorized memory location.

@harry-cpp : would this behaviour qualify as a bug? Thanks a lot.
